File: test/patchify.test.js
'use strict';

const test = require('node:test');
const assert = require('node:assert/strict');
const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');

const sharp = require('sharp');

const {
  PATCHIFY_DIRECTIVE_RE,
  parsePatchifyArgs,
  validatePatchifyArgs,
  computeGrid,
  patchifyImage,
  sourceLabelFor,
  DEFAULT_TILE_MIME_TYPE,
} = require('../src/patchify.js');
const { resolveIncludes } = require('../src/includes.js');
const { parseChatFile } = require('../src/parser.js');
const { parseFrontmatter } = require('../src/frontmatter.js');

function makeTmpDir() {
  return fs.mkdtempSync(path.join(os.tmpdir(), 'langchat-patchify-'));
}

async function makePng(width, height, channels = 3) {
  return await sharp({
    create: {
      width,
      height,
      channels,
      background: { r: 0xff, g: 0, b: 0 },
    },
  })
    .png()
    .toBuffer();
}

test('PATCHIFY_DIRECTIVE_RE matches the canonical form', () => {
  PATCHIFY_DIRECTIVE_RE.lastIndex = 0;
  const m = PATCHIFY_DIRECTIVE_RE.exec(
    '{{ patchify "Goku.png", 2, 2, 5, 10 }}'
  );
  assert.ok(m);
  assert.equal(m[1], 'Goku.png');
  assert.equal(m[2], '2');
  assert.equal(m[3], '2');
  assert.equal(m[4], '5');
  assert.equal(m[5], '10');
});

test('PATCHIFY_DIRECTIVE_RE tolerates extra whitespace', () => {
  PATCHIFY_DIRECTIVE_RE.lastIndex = 0;
  const variants = [
    '{{patchify "a.png",2,3,0.5,10}}',
    '{{  patchify  "b.png" , 1 , 4 , 99 , 0  }}',
    '{{\npatchify "c.png",\n3,\n3,\n25,\n25\n}}',
  ];
  for (const v of variants) {
    PATCHIFY_DIRECTIVE_RE.lastIndex = 0;
    assert.ok(PATCHIFY_DIRECTIVE_RE.exec(v), `should match: ${v}`);
  }
});

test('parsePatchifyArgs extracts integers from the directive text', () => {
  const out = parsePatchifyArgs('{{ patchify "a.png", 2, 3, 25, 0.5 }}');
  assert.deepEqual(out, {
    path: 'a.png',
    m: 2,
    n: 3,
    x: 25,
    y: 0.5,
  });
});

test('parsePatchifyArgs throws on malformed input', () => {
  assert.throws(
    () => parsePatchifyArgs('not a directive'),
    /cannot parse arguments/
  );
});

test('validatePatchifyArgs accepts the MVP5 example', () => {
  assert.doesNotThrow(() =>
    validatePatchifyArgs({ m: 2, n: 2, x: 5, y: 10 })
  );
});

test('validatePatchifyArgs rejects non-positive m and n', () => {
  assert.throws(() => validatePatchifyArgs({ m: 0, n: 2, x: 0, y: 0 }), /m/);
  assert.throws(() => validatePatchifyArgs({ m: -1, n: 2, x: 0, y: 0 }), /m/);
  assert.throws(() => validatePatchifyArgs({ m: 2, n: 0, x: 0, y: 0 }), /n/);
  assert.throws(
    () => validatePatchifyArgs({ m: 2.5, n: 2, x: 0, y: 0 }),
    /m must be a positive integer/
  );
});

test('validatePatchifyArgs rejects x and y outside [0, 100)', () => {
  assert.throws(
    () => validatePatchifyArgs({ m: 2, n: 2, x: -1, y: 0 }),
    /x \(vertical overlap/,
  );
  assert.throws(
    () => validatePatchifyArgs({ m: 2, n: 2, x: 0, y: -5 }),
    /y \(horizontal overlap/,
  );
  assert.throws(
    () => validatePatchifyArgs({ m: 2, n: 2, x: 100, y: 0 }),
    /x \(vertical overlap/,
  );
  assert.throws(
    () => validatePatchifyArgs({ m: 2, n: 2, x: 0, y: 100 }),
    /y \(horizontal overlap/,
  );
});

test('computeGrid produces m*n tiles in row-major order', () => {
  const grid = computeGrid(200, 100, 2, 3, 0, 0);
  assert.equal(grid.length, 6);
  for (let r = 0; r < 2; r++) {
    for (let c = 0; c < 3; c++) {
      const i = r * 3 + c;
      assert.equal(grid[i].row, r);
      assert.equal(grid[i].col, c);
    }
  }
});

test('computeGrid at 0% overlap produces a perfect tiling', () => {
  const grid = computeGrid(100, 80, 2, 2, 0, 0);
  assert.equal(grid.length, 4);
  for (const t of grid) {
    assert.equal(t.width, 50);
    assert.equal(t.height, 40);
  }
  assert.deepEqual(grid[0], { row: 0, col: 0, left: 0, top: 0, width: 50, height: 40 });
  assert.deepEqual(grid[1], { row: 0, col: 1, left: 50, top: 0, width: 50, height: 40 });
  assert.deepEqual(grid[2], { row: 1, col: 0, left: 0, top: 40, width: 50, height: 40 });
  assert.deepEqual(grid[3], { row: 1, col: 1, left: 50, top: 40, width: 50, height: 40 });
});

test('computeGrid with vertical overlap slides rows and bottom-aligns the last row', () => {
  const grid = computeGrid(100, 80, 2, 2, 50, 0);
  assert.equal(grid[0].top, 0);
  assert.equal(grid[1].top, 0);
  assert.equal(grid[2].top + grid[2].height, 80);
  assert.equal(grid[3].top + grid[3].height, 80);
});

test('computeGrid with horizontal overlap slides columns and right-aligns the last column', () => {
  const grid = computeGrid(100, 80, 2, 2, 0, 50);
  assert.equal(grid[0].left, 0);
  assert.equal(grid[2].left, 0);
  for (const r of [0, 1]) {
    assert.equal(grid[r * 2 + 1].left + grid[r * 2 + 1].width, 100);
  }
});

test('computeGrid with m=1 covers the full height regardless of x', () => {
  const grid = computeGrid(100, 80, 1, 4, 99, 0);
  assert.equal(grid.length, 4);
  for (const t of grid) {
    assert.equal(t.height, 80);
    assert.equal(t.top, 0);
  }
  assert.equal(grid[0].left, 0);
  assert.equal(grid[3].left + grid[3].width, 100);
});

test('computeGrid with n=1 covers the full width regardless of y', () => {
  const grid = computeGrid(100, 80, 4, 1, 0, 99);
  assert.equal(grid.length, 4);
  for (const t of grid) {
    assert.equal(t.width, 100);
    assert.equal(t.left, 0);
  }
});

test('sourceLabelFor produces "<base>[r<row>c<col>].png"', () => {
  assert.equal(sourceLabelFor('Goku.png', 0, 0), 'Goku[r0c0].png');
  assert.equal(sourceLabelFor('Goku.png', 1, 2), 'Goku[r1c2].png');
  assert.equal(sourceLabelFor('nested/path/a.bmp', 0, 0), 'a[r0c0].png');
  assert.equal(sourceLabelFor('noext', 2, 3), 'noext[r2c3].png');
});

test('patchifyImage with 1x1 produces one tile covering the whole image', async () => {
  const buf = await makePng(40, 30);
  const tiles = await patchifyImage(buf, { m: 1, n: 1, x: 5, y: 10 });
  assert.equal(tiles.length, 1);
  assert.equal(tiles[0].row, 0);
  assert.equal(tiles[0].col, 0);
  assert.equal(tiles[0].width, 40);
  assert.equal(tiles[0].height, 30);
  assert.equal(tiles[0].left, 0);
  assert.equal(tiles[0].top, 0);
  assert.equal(tiles[0].mimeType, DEFAULT_TILE_MIME_TYPE);
  const meta = await sharp(tiles[0].patch).metadata();
  assert.equal(meta.format, 'png');
  assert.equal(meta.width, 40);
  assert.equal(meta.height, 30);
});

test('patchifyImage produces m*n tiles re-encoded as PNG', async () => {
  const buf = await makePng(100, 80);
  const tiles = await patchifyImage(buf, { m: 2, n: 2, x: 5, y: 10 });
  assert.equal(tiles.length, 4);
  for (const t of tiles) {
    assert.equal(t.mimeType, 'image/png');
    const meta = await sharp(t.patch).metadata();
    assert.equal(meta.format, 'png');
    assert.ok(meta.width > 0);
    assert.ok(meta.height > 0);
  }
});

test('patchifyImage with 0% overlap partitions the image exactly', async () => {
  const buf = await makePng(100, 80);
  const tiles = await patchifyImage(buf, { m: 2, n: 2, x: 0, y: 0 });
  assert.equal(tiles.length, 4);
  for (const t of tiles) {
    assert.equal(t.width, 50);
    assert.equal(t.height, 40);
  }
  const byKey = (r, c) => tiles.find((t) => t.row === r && t.col === c);
  assert.equal(byKey(0, 0).left, 0);
  assert.equal(byKey(0, 0).top, 0);
  assert.equal(byKey(0, 1).left, 50);
  assert.equal(byKey(1, 0).top, 40);
  assert.equal(byKey(1, 1).left, 50);
  assert.equal(byKey(1, 1).top, 40);
});

test('patchifyImage re-encodes the source regardless of source mime', async () => {
  const png = await makePng(60, 40, 4);
  const tiles = await patchifyImage(png, { m: 2, n: 2, x: 0, y: 0 });
  for (const t of tiles) {
    assert.equal(t.mimeType, 'image/png');
    const meta = await sharp(t.patch).metadata();
    assert.equal(meta.format, 'png');
  }
});

test('resolveIncludes expands a patchify directive into placeholders + attachments', async () => {
  const dir = makeTmpDir();
  const imgPath = path.join(dir, 'a.png');
  fs.writeFileSync(imgPath, await makePng(100, 80));
  write(dir, 'chat.md', 'see {{ patchify "a.png", 2, 2, 0, 0 }} here');

  const result = await resolveIncludes(
    fs.readFileSync(path.join(dir, 'chat.md'), 'utf8'),
    { baseDir: dir }
  );

  assert.equal(
    result.text,
    'see {{ include "a[r0c0].png" }} {{ include "a[r0c1].png" }} {{ include "a[r1c0].png" }} {{ include "a[r1c1].png" }} here'
  );
  assert.equal(result.attachments.length, 4);
  for (const a of result.attachments) {
    assert.equal(a.type, 'image');
    assert.equal(a.mimeType, 'image/png');
  }
  assert.deepEqual(
    result.attachments.map((a) => a.source),
    ['a[r0c0].png', 'a[r0c1].png', 'a[r1c0].png', 'a[r1c1].png']
  );
});

test('resolveIncludes with patchify preserves directive order alongside text and image includes', async () => {
  const dir = makeTmpDir();
  const bg = await makePng(100, 80);
  const standalone = Buffer.from([0xff, 0xd8, 0xff, 0xd8]);
  fs.writeFileSync(path.join(dir, 'a.png'), bg);
  fs.writeFileSync(path.join(dir, 'b.jpg'), standalone);
  write(
    dir,
    'chat.md',
    '1:{{ include "b.jpg" }} 2:{{ patchify "a.png", 2, 2, 0, 0 }} 3:end'
  );

  const result = await resolveIncludes(
    fs.readFileSync(path.join(dir, 'chat.md'), 'utf8'),
    { baseDir: dir }
  );

  assert.equal(result.attachments.length, 5);
  assert.equal(result.attachments[0].source, 'b.jpg');
  assert.equal(result.attachments[1].source, 'a[r0c0].png');
  assert.equal(result.attachments[2].source, 'a[r0c1].png');
  assert.equal(result.attachments[3].source, 'a[r1c0].png');
  assert.equal(result.attachments[4].source, 'a[r1c1].png');

  assert.ok(result.text.includes('1:{{ include "b.jpg" }}'));
  assert.ok(result.text.includes('2:{{ include "a[r0c0].png" }}'));
  assert.ok(result.text.includes('3:end'));
});

test('resolveIncludes blocks path traversal on patchify', async () => {
  const dir = makeTmpDir();
  const outside = makeTmpDir();
  fs.writeFileSync(path.join(outside, 'secret.png'), await makePng(40, 40));
  write(dir, 'chat.md', '{{ patchify "../secret.png", 2, 2, 0, 0 }}');

  await assert.rejects(
    () =>
      resolveIncludes(
        fs.readFileSync(path.join(dir, 'chat.md'), 'utf8'),
        { baseDir: dir }
      ),
    /escapes base directory/
  );
});

test('resolveIncludes rejects invalid patchify arguments', async () => {
  const dir = makeTmpDir();
  fs.writeFileSync(path.join(dir, 'a.png'), await makePng(100, 80));
  write(dir, 'chat.md', '{{ patchify "a.png", 0, 2, 0, 0 }}');

  await assert.rejects(
    () =>
      resolveIncludes(
        fs.readFileSync(path.join(dir, 'chat.md'), 'utf8'),
        { baseDir: dir }
      ),
    /m must be a positive integer/
  );
});

test('resolveIncludes rejects x >= 100', async () => {
  const dir = makeTmpDir();
  fs.writeFileSync(path.join(dir, 'a.png'), await makePng(100, 80));
  write(dir, 'chat.md', '{{ patchify "a.png", 2, 2, 100, 0 }}');

  await assert.rejects(
    () =>
      resolveIncludes(
        fs.readFileSync(path.join(dir, 'chat.md'), 'utf8'),
        { baseDir: dir }
      ),
    /x \(vertical overlap/
  );
});

test('resolveIncludes rejects missing patchify source', async () => {
  const dir = makeTmpDir();
  write(dir, 'chat.md', '{{ patchify "nope.png", 2, 2, 0, 0 }}');

  await assert.rejects(
    () =>
      resolveIncludes(
        fs.readFileSync(path.join(dir, 'chat.md'), 'utf8'),
        { baseDir: dir }
      ),
    /patchify failed: nope\.png/
  );
});

test('resolveIncludes applies the size cap to patchify input', async () => {
  const dir = makeTmpDir();
  fs.writeFileSync(path.join(dir, 'big.png'), Buffer.alloc(64));
  write(dir, 'chat.md', '{{ patchify "big.png", 2, 2, 0, 0 }}');

  await assert.rejects(
    () =>
      resolveIncludes(
        fs.readFileSync(path.join(dir, 'chat.md'), 'utf8'),
        { baseDir: dir, maxBytes: 16 }
      ),
    /limit is 16 bytes/
  );
});

test('parser.js consumes patchify placeholders as image attachments in order', async () => {
  const dir = makeTmpDir();
  const bg = await makePng(100, 80);
  fs.writeFileSync(path.join(dir, 'a.png'), bg);

  const md = [
    '---',
    'model: qwen-vl-plus',
    '---',
    '',
    '# !user',
    '',
    'See {{ patchify "a.png", 2, 2, 0, 0 }} please',
    '',
  ].join('\n');

  const { body } = parseFrontmatter(md);
  const expanded = await resolveIncludes(body, { baseDir: dir });
  const { messages, outputSchema } = parseChatFile(
    expanded.text,
    expanded.attachments
  );
  assert.equal(outputSchema, null);
  assert.equal(messages.length, 1);
  const blocks = messages[0].content;
  assert.ok(Array.isArray(blocks));
  const images = blocks.filter((b) => b.type === 'image');
  assert.equal(images.length, 4);
  assert.deepEqual(
    images.map((b) => b.source),
    ['a[r0c0].png', 'a[r0c1].png', 'a[r1c0].png', 'a[r1c1].png']
  );
});

test('resolveIncludes handles a text-included file that itself uses patchify', async () => {
  const dir = makeTmpDir();
  fs.writeFileSync(path.join(dir, 'a.png'), await makePng(60, 40));
  write(dir, 'helper.md', 'P={{ patchify "a.png", 1, 2, 0, 0 }}');
  write(dir, 'chat.md', 'before {{ include "helper.md" }} after');

  const result = await resolveIncludes(
    fs.readFileSync(path.join(dir, 'chat.md'), 'utf8'),
    { baseDir: dir }
  );

  assert.equal(
    result.text,
    'before P={{ include "a[r0c0].png" }} {{ include "a[r0c1].png" }} after'
  );
  assert.equal(result.attachments.length, 2);
});

function write(dir, name, contents) {
  fs.writeFileSync(path.join(dir, name), contents);
}

File: go/internal/includes/patchify_test.go
package includes

import (
	"bytes"
	"image"
	"reflect"
	"regexp"
	"strings"
	"testing"
)

func TestPatchifyDirectiveREMatchesCanonical(t *testing.T) {
	m := PatchifyDirectiveRE.FindStringSubmatch(`{{ patchify "Goku.png", 2, 2, 5, 10 }}`)
	if m == nil {
		t.Fatalf("no match")
	}
	want := []string{
		`{{ patchify "Goku.png", 2, 2, 5, 10 }}`,
		"Goku.png", "2", "2", "5", "10",
	}
	if !reflect.DeepEqual(m, want) {
		t.Errorf("m = %v, want %v", m, want)
	}
}

func TestPatchifyDirectiveREToleratesWhitespace(t *testing.T) {
	variants := []string{
		`{{patchify "a.png",2,3,0.5,10}}`,
		`{{  patchify  "b.png" , 1 , 4 , 99 , 0  }}`,
		"{{\npatchify \"c.png\",\n3,\n3,\n25,\n25\n}}",
	}
	for _, v := range variants {
		if !PatchifyDirectiveRE.MatchString(v) {
			t.Errorf("did not match: %q", v)
		}
	}
}

func TestParsePatchifyArgs(t *testing.T) {
	out, err := ParsePatchifyArgs(`{{ patchify "a.png", 2, 3, 25, 0.5 }}`)
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	want := PatchifyArgs{Path: "a.png", M: 2, N: 3, X: 25, Y: 0.5}
	if out != want {
		t.Errorf("got %+v, want %+v", out, want)
	}
}

func TestParsePatchifyArgsMalformed(t *testing.T) {
	if _, err := ParsePatchifyArgs("not a directive"); err == nil {
		t.Errorf("expected error")
	} else if !strings.Contains(err.Error(), "cannot parse arguments") {
		t.Errorf("err = %v", err)
	}
}

func TestValidatePatchifyArgsMVP5Example(t *testing.T) {
	if err := ValidatePatchifyArgs(PatchifyArgs{M: 2, N: 2, X: 5, Y: 10}); err != nil {
		t.Errorf("err: %v", err)
	}
}

func TestValidatePatchifyArgsRejectsNonPositive(t *testing.T) {
	cases := []struct {
		args PatchifyArgs
		want string
	}{
		{PatchifyArgs{M: 0, N: 2, X: 0, Y: 0}, "m"},
		{PatchifyArgs{M: -1, N: 2, X: 0, Y: 0}, "m"},
		{PatchifyArgs{M: 2, N: 0, X: 0, Y: 0}, "n"},
	}
	for _, c := range cases {
		err := ValidatePatchifyArgs(c.args)
		if err == nil {
			t.Errorf("expected error for %+v", c.args)
			continue
		}
		if !strings.Contains(err.Error(), c.want) {
			t.Errorf("err %q does not contain %q", err.Error(), c.want)
		}
	}

	// Float-valued M is checked via the type system in Go; verify that a
	// value which would round-trip to non-integer M is still rejected. Since
	// the field is typed int we can't directly test "2.5"; instead verify
	// the integer-valued path accepts valid m.
	if err := ValidatePatchifyArgs(PatchifyArgs{M: 2, N: 2, X: 0, Y: 0}); err != nil {
		t.Errorf("m=2 should be valid: %v", err)
	}
}

func TestValidatePatchifyArgsRejectsXYOutOfRange(t *testing.T) {
	cases := []PatchifyArgs{
		{M: 2, N: 2, X: -1, Y: 0},
		{M: 2, N: 2, X: 0, Y: -5},
		{M: 2, N: 2, X: 100, Y: 0},
		{M: 2, N: 2, X: 0, Y: 100},
	}
	for _, c := range cases {
		if err := ValidatePatchifyArgs(c); err == nil {
			t.Errorf("expected error for %+v", c)
		}
	}
}

func TestComputeGridRowMajorOrder(t *testing.T) {
	grid := ComputeGrid(200, 100, 2, 3, 0, 0)
	if len(grid) != 6 {
		t.Fatalf("grid len = %d, want 6", len(grid))
	}
	for r := 0; r < 2; r++ {
		for c := 0; c < 3; c++ {
			i := r*3 + c
			if grid[i].Row != r || grid[i].Col != c {
				t.Errorf("grid[%d] = %+v, want row=%d col=%d", i, grid[i], r, c)
			}
		}
	}
}

func TestComputeGridZeroOverlapTiling(t *testing.T) {
	grid := ComputeGrid(100, 80, 2, 2, 0, 0)
	if len(grid) != 4 {
		t.Fatalf("grid len = %d, want 4", len(grid))
	}
	for _, t1 := range grid {
		if t1.Width != 50 || t1.Height != 40 {
			t.Errorf("tile = %+v, want 50x40", t1)
		}
	}
	want := []TileRect{
		{Row: 0, Col: 0, Left: 0, Top: 0, Width: 50, Height: 40},
		{Row: 0, Col: 1, Left: 50, Top: 0, Width: 50, Height: 40},
		{Row: 1, Col: 0, Left: 0, Top: 40, Width: 50, Height: 40},
		{Row: 1, Col: 1, Left: 50, Top: 40, Width: 50, Height: 40},
	}
	if !reflect.DeepEqual(grid, want) {
		t.Errorf("grid = %+v", grid)
	}
}

func TestComputeGridVerticalOverlapBottomAlign(t *testing.T) {
	grid := ComputeGrid(100, 80, 2, 2, 50, 0)
	if grid[0].Top != 0 || grid[1].Top != 0 {
		t.Errorf("top rows should be 0, got %d %d", grid[0].Top, grid[1].Top)
	}
	if grid[2].Top+grid[2].Height != 80 || grid[3].Top+grid[3].Height != 80 {
		t.Errorf("bottom rows should reach H=80, got %d + %d = %d", grid[2].Top, grid[2].Height, grid[2].Top+grid[2].Height)
	}
}

func TestComputeGridHorizontalOverlapRightAlign(t *testing.T) {
	grid := ComputeGrid(100, 80, 2, 2, 0, 50)
	if grid[0].Left != 0 || grid[2].Left != 0 {
		t.Errorf("left cols should be 0, got %d %d", grid[0].Left, grid[2].Left)
	}
	for _, r := range []int{0, 1} {
		i := r*2 + 1
		if grid[i].Left+grid[i].Width != 100 {
			t.Errorf("right col row %d: left+width = %d, want 100", r, grid[i].Left+grid[i].Width)
		}
	}
}

func TestComputeGridM1FullHeight(t *testing.T) {
	grid := ComputeGrid(100, 80, 1, 4, 99, 0)
	for _, t1 := range grid {
		if t1.Height != 80 || t1.Top != 0 {
			t.Errorf("tile = %+v, want height=80 top=0", t1)
		}
	}
	if grid[0].Left != 0 {
		t.Errorf("first tile left = %d, want 0", grid[0].Left)
	}
	if grid[3].Left+grid[3].Width != 100 {
		t.Errorf("last tile left+width = %d, want 100", grid[3].Left+grid[3].Width)
	}
}

func TestComputeGridN1FullWidth(t *testing.T) {
	grid := ComputeGrid(100, 80, 4, 1, 0, 99)
	for _, t1 := range grid {
		if t1.Width != 100 || t1.Left != 0 {
			t.Errorf("tile = %+v, want width=100 left=0", t1)
		}
	}
}

func TestSourceLabelFor(t *testing.T) {
	cases := []struct {
		path       string
		row, col   int
		want       string
	}{
		{"Goku.png", 0, 0, "Goku[r0c0].png"},
		{"Goku.png", 1, 2, "Goku[r1c2].png"},
		{"nested/path/a.bmp", 0, 0, "a[r0c0].png"},
		{"noext", 2, 3, "noext[r2c3].png"},
	}
	for _, c := range cases {
		got := SourceLabelFor(c.path, c.row, c.col)
		if got != c.want {
			t.Errorf("SourceLabelFor(%q, %d, %d) = %q, want %q", c.path, c.row, c.col, got, c.want)
		}
	}
}

// makePNG lives in includes_test.go and is shared with patchify_test.go.

func TestPatchifyImage1x1CoversWhole(t *testing.T) {
	buf := makePNG(t, 40, 30)
	tiles, err := Patchify(buf, PatchifyArgs{M: 1, N: 1, X: 5, Y: 10})
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if len(tiles) != 1 {
		t.Fatalf("len = %d, want 1", len(tiles))
	}
	t1 := tiles[0]
	if t1.Row != 0 || t1.Col != 0 || t1.Width != 40 || t1.Height != 30 || t1.Left != 0 || t1.Top != 0 {
		t.Errorf("tile = %+v", t1)
	}
	if t1.MIMEType != DefaultTileMIMEType {
		t.Errorf("mime = %q, want %q", t1.MIMEType, DefaultTileMIMEType)
	}
	meta, _, err := image.Decode(bytes.NewReader(t1.Patch))
	if err != nil {
		t.Fatalf("decode: %v", err)
	}
	if meta.Bounds().Dx() != 40 || meta.Bounds().Dy() != 30 {
		t.Errorf("decoded = %dx%d, want 40x30", meta.Bounds().Dx(), meta.Bounds().Dy())
	}
}

func TestPatchifyImageMNReencodedAsPNG(t *testing.T) {
	buf := makePNG(t, 100, 80)
	tiles, err := Patchify(buf, PatchifyArgs{M: 2, N: 2, X: 5, Y: 10})
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if len(tiles) != 4 {
		t.Fatalf("len = %d, want 4", len(tiles))
	}
	for _, t1 := range tiles {
		if t1.MIMEType != "image/png" {
			t.Errorf("mime = %q", t1.MIMEType)
		}
		if _, _, err := image.Decode(bytes.NewReader(t1.Patch)); err != nil {
			t.Errorf("tile not decodable: %v", err)
		}
	}
}

func TestPatchifyImageZeroOverlapPartitionsExactly(t *testing.T) {
	buf := makePNG(t, 100, 80)
	tiles, err := Patchify(buf, PatchifyArgs{M: 2, N: 2, X: 0, Y: 0})
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	for _, t1 := range tiles {
		if t1.Width != 50 || t1.Height != 40 {
			t.Errorf("tile = %+v", t1)
		}
	}
	byKey := func(r, c int) Tile {
		for _, t1 := range tiles {
			if t1.Row == r && t1.Col == c {
				return t1
			}
		}
		t.Fatalf("missing %d,%d", r, c)
		return Tile{}
	}
	if byKey(0, 0).Left != 0 || byKey(0, 0).Top != 0 {
		t.Errorf("(0,0) = %+v", byKey(0, 0))
	}
	if byKey(0, 1).Left != 50 {
		t.Errorf("(0,1) Left = %d", byKey(0, 1).Left)
	}
	if byKey(1, 0).Top != 40 {
		t.Errorf("(1,0) Top = %d", byKey(1, 0).Top)
	}
	if byKey(1, 1).Left != 50 || byKey(1, 1).Top != 40 {
		t.Errorf("(1,1) = %+v", byKey(1, 1))
	}
}

func TestPatchifyRegexIsAccessible(t *testing.T) {
	// Just a sanity check that the regex variable is exported.
	m := PatchifyDirectiveRE.MatchString(`x {{ patchify "a.png", 1, 1, 0, 0 }} y`)
	if !m {
		t.Errorf("expected match")
	}
	if regexp.MustCompile(PatchifyDirectiveRE.String()).NumSubexp() != 5 {
		t.Errorf("expected 5 capture groups")
	}
}